# HG changeset patch
# Parent  1c4504c4db4fe53a7ae2e2f3787d8d0e6ef0c46c

diff --git a/nsis/gvim.nsi b/nsis/gvim.nsi
--- a/nsis/gvim.nsi
+++ b/nsis/gvim.nsi
@@ -42,10 +42,16 @@
 !include LogicLib.nsh
 !include x64.nsh
 
-Name "Vim ${VER_MAJOR}.${VER_MINOR}"
+!define PRODUCT		"Vim ${VER_MAJOR}.${VER_MINOR}"
+!define UNINST_REG_KEY	"Software\Microsoft\Windows\CurrentVersion\Uninstall\${PRODUCT}"
+!define PUBLISHER	"Bram Moolenaar et al."
+
+Name "${PRODUCT}"
 OutFile gvim${VER_MAJOR}${VER_MINOR}.exe
 CRCCheck force
 SetCompressor /SOLID lzma
+SetCompressorDictSize 64
+ManifestDPIAware true
 SetDatablockOptimize on
 RequestExecutionLevel highest
 XPStyle on
@@ -195,7 +201,7 @@ Function un.GetParent
 FunctionEnd
 
 ##########################################################
-Section "Vim executables and runtime files"
+Section "Vim executables and runtime files" sec_main_id
 	SectionIn 1 2 3 RO
 
 	# we need also this here if the user changes the instdir
@@ -299,7 +305,7 @@ Section "Vim executables and runtime fil
 SectionEnd
 
 ##########################################################
-Section "Vim console program (vim.exe)"
+Section "Vim console program (vim.exe)" sec_cons_id
 	SectionIn 1 3
 
 	SetOutPath $0
@@ -400,7 +406,7 @@ Section "Create plugin directories in VI
 SectionEnd
 
 ##########################################################
-Section "VisVim Extension for MS Visual Studio"
+Section "VisVim Extension for MS Visual Studio" sec_visvim_id
 	SectionIn 3
 
 	SetOutPath $0
@@ -410,7 +416,7 @@ SectionEnd
 
 ##########################################################
 !ifdef HAVE_NLS
-Section "Native Language Support"
+Section "Native Language Support" sec_nls_id
 	SectionIn 1 3
 
 	SetOutPath $0\lang
@@ -489,6 +495,39 @@ SectionEnd
 
 ##########################################################
 Section -post
+
+	# Get estimated install size
+	SectionGetSize ${sec_main_id} $3
+	${If} ${SectionIsSelected} ${sec_cons_id}
+	  SectionGetSize ${sec_cons_id} $4
+	  IntOp $3 $3 + $4
+	${EndIf}
+	${If} ${SectionIsSelected} ${sec_gvimext_id}
+	  SectionGetSize ${sec_gvimext_id} $4
+	  IntOp $3 $3 + $4
+	${EndIf}
+	${If} ${SectionIsSelected} ${sec_visvim_id}
+	  SectionGetSize ${sec_visvim_id} $4
+	  IntOp $3 $3 + $4
+	${EndIf}
+	${If} ${SectionIsSelected} ${sec_nls_id}
+	  SectionGetSize ${sec_nls_id} $4
+	  IntOp $3 $3 + $4
+	${EndIf}
+
+	# Register some information.
+	# Other information will be set by the install.exe.
+	${If} ${RunningX64}
+	  SetRegView 64
+	${EndIf}
+	WriteRegStr HKLM "${UNINST_REG_KEY}" "DisplayIcon" '"$0\gvim.exe",0'
+	WriteRegStr HKLM "${UNINST_REG_KEY}" "DisplayVersion" "${VER_MAJOR}.${VER_MINOR}"
+	WriteRegStr HKLM "${UNINST_REG_KEY}" "Publisher" "${PUBLISHER}"
+	WriteRegDWORD HKLM "${UNINST_REG_KEY}" "EstimatedSize" $3
+	${If} ${RunningX64}
+	  SetRegView lastused
+	${EndIf}
+
 	BringToFront
 SectionEnd
 
